Fix the possible buffer overrun reported by Roland Kay in xine-lib's issue #1603458.
Patch by JW in issue #1603470.

See Gentoo bug #156645.
Index: xine-lib-1.1.2/src/input/libreal/asmrp.c
===================================================================
--- xine-lib-1.1.2.orig/src/input/libreal/asmrp.c
+++ xine-lib-1.1.2/src/input/libreal/asmrp.c
@@ -604,7 +604,7 @@ static int asmrp_rule (asmrp_t *p) {
   return ret;
 }
 
-static int asmrp_eval (asmrp_t *p, int *matches) {
+static int asmrp_eval (asmrp_t *p, int *matches, int matchsize) {
 
   int rule_num, num_matches;
 
@@ -613,7 +613,7 @@ static int asmrp_eval (asmrp_t *p, int *
   asmrp_get_sym (p);
 
   rule_num = 0; num_matches = 0;
-  while (p->sym != ASMRP_SYM_EOF) {
+  while (p->sym != ASMRP_SYM_EOF && num_matches < matchsize - 1) {
 
     if (asmrp_rule (p)) {
       lprintf ("rule #%d is true\n", rule_num);
@@ -629,7 +629,7 @@ static int asmrp_eval (asmrp_t *p, int *
   return num_matches;
 }
 
-int asmrp_match (const char *rules, int bandwidth, int *matches) {
+int asmrp_match (const char *rules, int bandwidth, int *matches, int matchsize) {
 
   asmrp_t *p;
   int      num_matches;
@@ -641,7 +641,7 @@ int asmrp_match (const char *rules, int 
   asmrp_set_id (p, "Bandwidth", bandwidth);
   asmrp_set_id (p, "OldPNMPlayer", 0);
 
-  num_matches = asmrp_eval (p, matches);
+  num_matches = asmrp_eval (p, matches, matchsize);
 
   asmrp_dispose (p);
 
Index: xine-lib-1.1.2/src/input/libreal/asmrp.h
===================================================================
--- xine-lib-1.1.2.orig/src/input/libreal/asmrp.h
+++ xine-lib-1.1.2/src/input/libreal/asmrp.h
@@ -37,6 +37,6 @@
 #ifndef HAVE_ASMRP_H
 #define HAVE_ASMRP_H
 
-int asmrp_match (const char *rules, int bandwidth, int *matches) ;
+int asmrp_match (const char *rules, int bandwidth, int *matches, int matchesizxe) ;
 
 #endif
Index: xine-lib-1.1.2/src/input/libreal/real.c
===================================================================
--- xine-lib-1.1.2.orig/src/input/libreal/real.c
+++ xine-lib-1.1.2/src/input/libreal/real.c
@@ -476,7 +476,7 @@ rmff_header_t *real_parse_sdp(char *data
 
     lprintf("calling asmrp_match with:\n%s\n%u\n", desc->stream[i]->asm_rule_book, bandwidth);
 
-    n=asmrp_match(desc->stream[i]->asm_rule_book, bandwidth, rulematches);
+    n=asmrp_match(desc->stream[i]->asm_rule_book, bandwidth, rulematches, sizeof(rulematches)/sizeof(rulematches[0]));
     for (j=0; j<n; j++) {
       lprintf("asmrp rule match: %u for stream %u\n", rulematches[j], desc->stream[i]->stream_id);
       sprintf(b,"stream=%u;rule=%u,", desc->stream[i]->stream_id, rulematches[j]);
